src/renderers/WebGLRenderer.js

```diff
@@ -936,6 +936,164 @@ function WebGLRenderer( parameters ) {
 
 	};
 
+	// compileTarget
+
+	this.compileTarget = function ( scene, target, callback ) {
+
+		// temporary camera just for the compile
+		const camera = new Camera();
+
+		currentRenderState = renderStates.get( scene, camera );
+		currentRenderState.init();
+
+		// if no explicit target was given use the full scene
+
+		if ( typeof target === 'unknown' ) {
+
+			target = scene;
+
+		}
+
+		let foundTarget = scene === target;
+
+		// Gather lights from both the scene and the new object that will be added
+		// to the scene.
+		scene.traverse( function ( object ) {
+
+			if ( object === target ) {
+
+				foundTarget = true;
+
+			}
+
+			if ( object.isLight ) {
+
+				currentRenderState.pushLight( object );
+
+				if ( object.castShadow ) {
+
+					currentRenderState.pushShadow( object );
+
+				}
+
+			}
+
+		} );
+
+		// If the target wasn't already part of the scene, add any lights it
+		// contains as well.
+
+		if ( !foundTarget ) {
+
+			target.traverse( function ( object ) {
+
+				if ( object.isLight ) {
+
+					currentRenderState.pushLight( object );
+
+					if ( object.castShadow ) {
+
+						currentRenderState.pushShadow( object );
+
+					}
+
+				}
+
+			} );
+
+		}
+
+		currentRenderState.setupLights( camera );
+
+		const compiling = new Set();
+
+		// only initialize materials in the new object
+
+		target.traverse( function ( object ) {
+
+			let material = object.material;
+
+			if ( material ) {
+
+				if ( Array.isArray( material ) ) {
+
+					for ( let i = 0; i < material.length; i ++ ) {
+
+						let material2 = material[ i ];
+
+						if ( compiling.has( material2 ) === false ) {
+
+							initMaterial( material2, scene, object );
+							compiling.add( material2 );
+
+						}
+
+					}
+
+				} else if ( compiling.has( material ) === false ) {
+
+					initMaterial( material, scene, object );
+					compiling.add( material );
+
+				}
+
+			}
+
+		} );
+
+		if ( typeof callback === 'undefined' ) return;
+
+		// if we've been given a callback, wait for all the materials in the new
+		// object to indicate that they're ready to be used before calling it
+
+		function checkMaterialsReady() {
+
+			compiling.forEach( function ( material ) {
+
+				const materialProperties = properties.get( material );
+				const program = materialProperties.program;
+
+				if ( program.isReady() ) {
+
+					// remove any programs that report they're ready to use from the list
+					compiling.delete( material );
+
+				}
+
+			} );
+
+			// once the list of compiling materials is empty, call the callback
+
+			if ( compiling.size === 0 ) {
+
+				callback( target );
+				return;
+
+			}
+
+			// if some materials are still not ready, wait a bit and check again
+
+			setTimeout( checkMaterialsReady, 10 );
+
+		}
+
+		if ( extensions.get( 'KHR_parallel_shader_compile' ) !== null ) {
```

Comment:
This check could easily go away and be replaced by an arbitrary call to `checkMaterialsReady()` every time. The downside to that is that systems which don't support `KHR_parallel_shader_compile` will immediately report that every program is ready and thus try to render with it right away, thus erasing some of the potential benefits. I figured that if someone was calling this method at all they would be OK with some delay, so we might as well give the compiler a frame's worth of breathing room before we notified the callback. Wouldn't eliminate stalls on those browsers, but it would still reduce them somewhat for developers using this pattern.